### PR TITLE
Port X11 backend to the EVL2.0 API

### LINKED ISSUE #266: EventsLoopProxy does not wakeup the EventsLoop unless it's already blocked
See the conrod issue for a discussion on this:

https://github.com/PistonDevelopers/conrod/issues/1051

To get a non-polling multi-threaded app running the conrod examples do the following:

  1. Create a rendering thread that reads inputs and writes it's output to a set of `std::sync::mpsc` channels
  2. Block the main loop on `EventsLoop::run_forever()`
  3. In the rendering thread whenever we have written the output use an `EventsLoopProxy::wakeup()` to wake up the main thread
  4. Inside `run_forever()` pass on events to the channel until we get an `Awakened`
  5. Finish the main loop by rendering from the second channel and go back to blocking in `run_forever()`

The problem with this setup is that step 3 doesn't reliably unblock `run_forever()` because if you use `EventsLoopProxy::wakeup()` when not inside `run_forever()` the next time `run_forever()` runs it will not immediately unblock.

Is this intended behavior? @mitchmindtree was expecting the semantics of `wakeup()` to be that if it's called it will unblock `run_forever()` immediately when it gets called and that does sound much more useful.

### PR DESCRIPTION
**Pull request by vberger**
**This code has only been lightly test and I'd love if a few persons could check it locally and test it more.**

This ports the X11 backend  to the EVL2.0 API. To have an easier time implementing it I used [calloop](https://docs.rs/calloop/) to drive the event loop:

- This is not a new dependency (the wayland backend already uses it)
- This means events readiness is handled directly by calloop, so there is no longer need for a dummy X11 window to wakeup a blocking call waiting for events
- I moved the event-processing logic into a new file, because `mod.rs` was really huge (which is a large part of the diff)

Fixes #266

### CODE REVIEW

**Comment by ZeGentzy** [comment]
Not generating close requested messages, when I press Meta-Shift-Q. I use i3wm, fwiw.

**Comment by vberger** [comment]
Ah right, I missed this line of the `ControlFlow` documentation :

> Almost every change is persistent between multiple calls to the event loop closure within a given run loop. The only exception to this is `Exit` which, once set, cannot be unset.

In that case the wayland backend has the same bug, I'll correct that.

**Comment by mitchmindtree** [comment]
@ZeGentzy can you confirm this did not happen previously? I know that previously in debug mode one of the examples used to print all events which would flood I/O when you moved the mouse a lot and cause close events to delay, but I thought this was just due to the huge amount of events being printed without rustc optimisations enabled. I'm not sure you're running into the same thing though, I'm currently on potato internet in australia and can't buffer the video.

**Comment by ZeGentzy** [comment]
@mitchmindtree This did not happen to me previously. This still happens even if I comment out all the `println`s.

**Comment by vberger** [comment]
While I kinda understand why it could happen, I don't manage to reproduce it on my system.

My guess is that you manage to generate events faster than winit consumes them. Given that the event loop structure is to process all pending events before sending an `EventsCleared`, and the actual exiting is done after `EventsCleared`. But from my understanding of the evl2 API this is intended behavior.

The real question is how do you generate events faster than winit can process them? Given the processing of events is made in two steps, the `println!()` in the examples are unrelated to that.

- Do you have some high-resolution mouse that generates a huge amount of events?
- Does this delay in closing still happen with the examples built in release mode?

**Comment by ZeGentzy** [comment]
>     * Do you have some high-resolution mouse that generates a huge amount of events?

Afaik, no, I just got one of these:
https://www.microsoft.com/accessories/en-us/products/mice/basic-optical-mouse/p58-00061

>     * Does this delay in closing still happen with the examples built in release mode?

I'm afraid it still happens.

**Comment by ZeGentzy** [comment]
It appears the new X11 backend is just a lot slower. 

Compare the old one:
```
[2019-04-19T20:22:52.319262835Z WARN  window] DeviceEvent { device_id: DeviceId(X(DeviceId(9))), event: Motion { axis: 0, value: -1.0 } }
[2019-04-19T20:22:52.319374713Z WARN  window] DeviceEvent { device_id: DeviceId(X(DeviceId(9))), event: Motion { axis: 1, value: -1.0 } }
[2019-04-19T20:22:52.319423524Z WARN  window] DeviceEvent { device_id: DeviceId(X(DeviceId(9))), event: MouseMotion { delta: (-1.0, -1.0) } }
[2019-04-19T20:22:52.319536131Z WARN  window] WindowEvent { window_id: WindowId(X(WindowId(44040195))), event: CursorMoved { device_id: DeviceId(X(DeviceId(2))), position: LogicalPosition { x: 900.6574894831731, y: 335.5100661057692 }, modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } }
[2019-04-19T20:22:52.319657799Z WARN  window] WindowEvent { window_id: WindowId(X(WindowId(44040195))), event: AxisMotion { device_id: DeviceId(X(DeviceId(2))), axis: 0, value: 977.7122805239633 } }
[2019-04-19T20:22:52.319713775Z WARN  window] WindowEvent { window_id: WindowId(X(WindowId(44040195))), event: AxisMotion { device_id: DeviceId(X(DeviceId(2))), axis: 1, value: 387.46924707107246 } }
[2019-04-19T20:22:52.319772767Z WARN  window] DeviceEvent { device_id: DeviceId(X(DeviceId(9))), event: Motion { axis: 0, value: -1.0 } }
[2019-04-19T20:22:52.319816991Z WARN  window] DeviceEvent { device_id: DeviceId(X(DeviceId(9))), event: Motion { axis: 1, value: -1.0 } }
[2019-04-19T20:22:52.319866292Z WARN  window] DeviceEvent { device_id: DeviceId(X(DeviceId(9))), event: MouseMotion { delta: (-1.0, -1.0) } }
[2019-04-19T20:22:52.319922156Z WARN  window] WindowEvent { window_id: WindowId(X(WindowId(44040195))), event: CursorMoved { device_id: DeviceId(X(DeviceId(2))), position: LogicalPosition { x: 899.7344125600962, y: 334.5869891826923 }, modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } }
[2019-04-19T20:22:52.320011536Z WARN  window] WindowEvent { window_id: WindowId(X(WindowId(44040195))), event: AxisMotion { device_id: DeviceId(X(DeviceId(2))), axis: 0, value: 976.7122805239633 } }
[2019-04-19T20:22:52.320066465Z WARN  window] WindowEvent { window_id: WindowId(X(WindowId(44040195))), event: AxisMotion { device_id: DeviceId(X(DeviceId(2))), axis: 1, value: 386.46924707107246 } }

[2019-04-19T20:22:52.337818009Z WARN  window] DeviceEvent { device_id: DeviceId(X(DeviceId(9))), event: Motion { axis: 0, value: -1.0 } }
[2019-04-19T20:22:52.338274721Z WARN  window] DeviceEvent { device_id: DeviceId(X(DeviceId(9))), event: Motion { axis: 1, value: -1.0 } }
[2019-04-19T20:22:52.338376105Z WARN  window] DeviceEvent { device_id: DeviceId(X(DeviceId(9))), event: MouseMotion { delta: (-1.0, -1.0) } }
[2019-04-19T20:22:52.338606807Z WARN  window] WindowEvent { window_id: WindowId(X(WindowId(44040195))), event: CursorMoved { device_id: DeviceId(X(DeviceId(2))), position: LogicalPosition { x: 898.8113356370193, y: 333.6639122596154 }, modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } }
[2019-04-19T20:22:52.338864167Z WARN  window] WindowEvent { window_id: WindowId(X(WindowId(44040195))), event: AxisMotion { device_id: DeviceId(X(DeviceId(2))), axis: 0, value: 975.7122805239633 } }
[2019-04-19T20:22:52.338968633Z WARN  window] WindowEvent { window_id: WindowId(X(WindowId(44040195))), event: AxisMotion { device_id: DeviceId(X(DeviceId(2))), axis: 1, value: 385.46924707107246 } }
[2019-04-19T20:22:52.339081204Z WARN  window] DeviceEvent { device_id: DeviceId(X(DeviceId(9))), event: Motion { axis: 0, value: -2.0 } }
[2019-04-19T20:22:52.339165588Z WARN  window] DeviceEvent { device_id: DeviceId(X(DeviceId(9))), event: Motion { axis: 1, value: -2.0 } }
[2019-04-19T20:22:52.339257750Z WARN  window] DeviceEvent { device_id: DeviceId(X(DeviceId(9))), event: MouseMotion { delta: (-2.0, -2.0) } }
[2019-04-19T20:22:52.339364173Z WARN  window] WindowEvent { window_id: WindowId(X(WindowId(44040195))), event: CursorMoved { device_id: DeviceId(X(DeviceId(2))), position: LogicalPosition { x: 896.9651817908655, y: 331.81775841346155 }, modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } }
[2019-04-19T20:22:52.339534605Z WARN  window] WindowEvent { window_id: WindowId(X(WindowId(44040195))), event: AxisMotion { device_id: DeviceId(X(DeviceId(2))), axis: 0, value: 973.7122805239633 } }
[2019-04-19T20:22:52.339637447Z WARN  window] WindowEvent { window_id: WindowId(X(WindowId(44040195))), event: AxisMotion { device_id: DeviceId(X(DeviceId(2))), axis: 1, value: 383.46924707107246 } }
```

To an equivalent chunk of the new backend, which is a lot slower:
```
[2019-04-19T20:20:15.793298779Z WARN  window] DeviceEvent { device_id: DeviceId(X(DeviceId(9))), event: Motion { axis: 0, value: -5.0 } }
[2019-04-19T20:20:15.809188850Z WARN  window] DeviceEvent { device_id: DeviceId(X(DeviceId(9))), event: Motion { axis: 1, value: 3.0 } }
[2019-04-19T20:20:15.825503716Z WARN  window] DeviceEvent { device_id: DeviceId(X(DeviceId(9))), event: MouseMotion { delta: (-5.0, 3.0) } }
[2019-04-19T20:20:15.841550367Z WARN  window] WindowEvent { window_id: WindowId(X(WindowId(41943042))), event: CursorMoved { device_id: DeviceId(X(DeviceId(2))), position: LogicalPosition { x: 1574.4476412259617, y: 169.92748084435098 }, modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } }
[2019-04-19T20:20:15.863795344Z WARN  window] WindowEvent { window_id: WindowId(X(WindowId(41943042))), event: AxisMotion { device_id: DeviceId(X(DeviceId(2))), axis: 0, value: 1707.6515653505921 } }
[2019-04-19T20:20:15.876424096Z WARN  window] WindowEvent { window_id: WindowId(X(WindowId(41943042))), event: AxisMotion { device_id: DeviceId(X(DeviceId(2))), axis: 1, value: 208.0881081011612 } }
[2019-04-19T20:20:15.891262484Z WARN  window] DeviceEvent { device_id: DeviceId(X(DeviceId(9))), event: Motion { axis: 0, value: -1.0 } }
[2019-04-19T20:20:15.908223124Z WARN  window] DeviceEvent { device_id: DeviceId(X(DeviceId(9))), event: Motion { axis: 1, value: 1.0 } }
[2019-04-19T20:20:15.924468019Z WARN  window] DeviceEvent { device_id: DeviceId(X(DeviceId(9))), event: MouseMotion { delta: (-1.0, 1.0) } }
[2019-04-19T20:20:15.942122705Z WARN  window] WindowEvent { window_id: WindowId(X(WindowId(41943042))), event: CursorMoved { device_id: DeviceId(X(DeviceId(2))), position: LogicalPosition { x: 1573.5245643028848, y: 170.8505577674279 }, modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } }
[2019-04-19T20:20:15.957567987Z WARN  window] WindowEvent { window_id: WindowId(X(WindowId(41943042))), event: AxisMotion { device_id: DeviceId(X(DeviceId(2))), axis: 0, value: 1706.6515653505921 } }
[2019-04-19T20:20:15.977592841Z WARN  window] WindowEvent { window_id: WindowId(X(WindowId(41943042))), event: AxisMotion { device_id: DeviceId(X(DeviceId(2))), axis: 1, value: 209.0881081011612 } }

[2019-04-19T20:20:15.994205505Z WARN  window] DeviceEvent { device_id: DeviceId(X(DeviceId(9))), event: Motion { axis: 0, value: -3.0 } }
[2019-04-19T20:20:16.009690888Z WARN  window] DeviceEvent { device_id: DeviceId(X(DeviceId(9))), event: Motion { axis: 1, value: 3.0 } }
[2019-04-19T20:20:16.025523668Z WARN  window] DeviceEvent { device_id: DeviceId(X(DeviceId(9))), event: MouseMotion { delta: (-3.0, 3.0) } }
[2019-04-19T20:20:16.043194106Z WARN  window] WindowEvent { window_id: WindowId(X(WindowId(41943042))), event: CursorMoved { device_id: DeviceId(X(DeviceId(2))), position: LogicalPosition { x: 1570.689415564904, y: 173.68562199519232 }, modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } }
[2019-04-19T20:20:16.057976298Z WARN  window] WindowEvent { window_id: WindowId(X(WindowId(41943042))), event: AxisMotion { device_id: DeviceId(X(DeviceId(2))), axis: 0, value: 1703.5802479502745 } }
[2019-04-19T20:20:16.075623047Z WARN  window] WindowEvent { window_id: WindowId(X(WindowId(41943042))), event: AxisMotion { device_id: DeviceId(X(DeviceId(2))), axis: 1, value: 212.15942550171167 } }
[2019-04-19T20:20:16.092335668Z WARN  window] DeviceEvent { device_id: DeviceId(X(DeviceId(9))), event: Motion { axis: 0, value: -2.0 } }
[2019-04-19T20:20:16.108555042Z WARN  window] DeviceEvent { device_id: DeviceId(X(DeviceId(9))), event: Motion { axis: 1, value: 2.0 } }
[2019-04-19T20:20:16.128534013Z WARN  window] DeviceEvent { device_id: DeviceId(X(DeviceId(9))), event: MouseMotion { delta: (-2.0, 2.0) } }
[2019-04-19T20:20:16.141237170Z WARN  window] WindowEvent { window_id: WindowId(X(WindowId(41943042))), event: CursorMoved { device_id: DeviceId(X(DeviceId(2))), position: LogicalPosition { x: 1568.7772310697117, y: 175.5978769155649 }, modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } }
[2019-04-19T20:20:16.158525995Z WARN  window] WindowEvent { window_id: WindowId(X(WindowId(41943042))), event: AxisMotion { device_id: DeviceId(X(DeviceId(2))), axis: 0, value: 1701.5086335253436 } }
[2019-04-19T20:20:16.174090090Z WARN  window] WindowEvent { window_id: WindowId(X(WindowId(41943042))), event: AxisMotion { device_id: DeviceId(X(DeviceId(2))), axis: 1, value: 214.23103992664255 } }
```
Further investigation reveals this:
```
2019-04-19T20:52:50.918839701Z WARN  winit::platform_impl::platform::x11] gentz: pending events callback start
[2019-04-19T20:52:50.918873070Z WARN  window] Callback at start DeviceEvent { device_id: DeviceId(X(DeviceId(9))), event: Motion { axis: 0, value: -23.0 } }
[2019-04-19T20:52:50.918923701Z WARN  window] Callback post match 1 DeviceEvent { device_id: DeviceId(X(DeviceId(9))), event: Motion { axis: 0, value: -23.0 } }
[2019-04-19T20:52:50.935596215Z WARN  window] Callback post swap DeviceEvent { device_id: DeviceId(X(DeviceId(9))), event: Motion { axis: 0, value: -23.0 } }
[2019-04-19T20:52:50.935728835Z WARN  window] Callback post match 2 DeviceEvent { device_id: DeviceId(X(DeviceId(9))), event: Motion { axis: 0, value: -23.0 } }
[2019-04-19T20:52:50.935779131Z WARN  winit::platform_impl::platform::x11] gentz: pending events callback end
```

Indeed, glutin is swaping it's buffers for every mouse moved event, which is greatly slowing things down. Shouldn't we only be redrawing on some redraw event? How is glutin mean't to handle this?

File: src/platform_impl/linux/wayland/event_loop.rs [hunk 1]
```diff
@@ -204,23 +204,63 @@ impl<T: 'static> EventLoop<T> {
             // empty buffer of events
             {
                 let mut guard = sink.lock().unwrap();
-                guard.empty_with(|evt| callback(evt, &self.window_target, &mut control_flow));
+                guard.empty_with(|evt| {
+                    // make ControlFlow::Exit sticky by providing a dummy
+                    // control flow reference if it is already Exit.
+                    let mut dummy = ControlFlow::Exit;
```
Comment (ZeGentzy) [inline src/platform_impl/linux/wayland/event_loop.rs]: Can we avoid the code duplication by separating this out in a function. I wrote a small patch: https://termbin.com/lq9d

**Comment by ZeGentzy** [comment]
Oh, I guess there is a `RedrawRequested` `WindowEvent`. Using that fixes that.

**Comment by vberger** [comment]
> Oh, I guess there is a `RedrawRequested` `WindowEvent`. Using that fixes that.

Ah yes, forgot about that, this is its intended use I believe.

I have a few questions wrt to `RedrawRequested` though @Osspial:

- How would apps that need a redraw in sync with the monitor use it, should winit provide a way for the app to request receiving such an event at regular intervals matching vsync?
- Should several user-triggered requests for redraw be collapsed into a single `RedrawRequested`, so that this even can occur at most once in an eventloop cycle?

More generally, how does `RedrawRequested` fit exactly into the "process input - redraw" loop of games?

**Comment by Osspial** [comment]
> How would apps that need a redraw in sync with the monitor use it, should winit provide a way for the app to request receiving such an event at regular intervals matching vsync?

I've always figured that vsync could be handled sufficiently by calling `swap_buffers` with vsync enabled. If that isn't the case we could look at an API solution, but if there's some way to make that work it would be simpler.

> Should several user-triggered requests for redraw be collapsed into a single `RedrawRequested`, so that this even can occur at most once in an eventloop cycle?

Almost - all redraw requests from both the user and the OS should be collapsed per-window as well as per-cycle, so only a single `RedrawRequested` event can be generated for each window in each cycle.

Also, `RedrawRequested` ideally should always be the last event generated in a cycle, if that's possible to do. That guarantees that all input events have been processed before redrawing is performed. Win32 does this automatically, but I'm not sure how feasible it is to do on other OSes.

**Comment by Ralith** [comment]
> Indeed, glutin is swaping it's buffers for every mouse moved event, which is greatly slowing things down. Shouldn't we only be redrawing on some redraw event? How is glutin mean't to handle this?

This is a common mistake. You should always *fully* drain the event queue before doing any further processing (e.g. rendering).

**Comment by vberger** [comment]
@Osspial Thanks for the clarification, I'll need to rework the wayland & x11 backends to match this then. I agree this should be detailed in some documentation.

Maybe we should have two basic setups presented: a "game-like" app that draws with vsync and uses `Poll`, and a "gui-like" app that only draws when some of its content changes and uses `Wait` or `WaitTimeout`.

**Comment by vberger** [comment]
I guess this PR should be good for merging now, is there anything missing?

**Comment by vberger** [comment]
Ah right, forgot about the code-duplication issue. Will correct that.

**Comment by vberger** [comment]
Okay, I'll merge this so that it can be tested more easily, and any needed future bugfixes will be a subsequent PRs.

**Pull request merged**